kokoro/unittests/dart_test_json2xml.py

```diff
@@ -0,0 +1,389 @@
+#!/usr/bin/env python3
+
+from collections import defaultdict
+from datetime import datetime, timedelta, timezone
+from enum import IntEnum
+import fileinput
+import json
+from typing import Iterable
+from lxml import etree
+import os
+import sys
+
+
+class TestElement:
+  _ISO8601_FORMAT = '%Y-%m-%dT%H:%M:%SZ'
+
+  def __init__(self, info: dict) -> None:
+    self._attrs = {}
+    self._properties = {}
+    self._unprocessed = info
+    self.end_time = 0  # ms since start
+
+  def pop_info(self, key: str, default=None):
+    return self._unprocessed.pop(key, default)
+
+  def update_properties_with_unprocessed(self) -> None:
+    self._properties.update(
+        (k, v) for k, v in self._unprocessed.items() if v is not None)
+    self._unprocessed = {}
+
+  def update_end_time(self, new_time):
+    self.end_time = max(self.end_time, new_time)
+
+  def compute_timings(self, time_zero):
+    ms_since_start = self.pop_info('time')
+    assert ms_since_start is not None
+    start_time = time_zero + timedelta(milliseconds=ms_since_start)
+    self._attrs['timestamp'] = start_time.strftime(self._ISO8601_FORMAT)
+    self._attrs['time'] = '{:.3f}'.format(
+        (self.end_time - ms_since_start) / 1000)
+
+  def finalize_attrs(self) -> None:
```

Comment:
Finalize seems to really be to_string. Something like that would be a more descriptive name.

Reply:
Thanks! I have changed it into `convert_attrs_to_str`